iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+import polars as pl
+import datetime as dt
+from sklearn.linear_model import LinearRegression
+import numpy as np
+import abc
+from typing_extensions import Self
+
+
+class UptakeData(pl.DataFrame, metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def validate(self) -> None:
+        pass
+
+
+class IncidentUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "season", "elapsed", "interval", "estimate", "previous"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert all(
+            self[x].dtype == pl.Utf8 for x in ["region", "season"]
+        ), "Column 'region' and/or 'season' is not a string."
+        assert all(
+            self[x].dtype.is_numeric()
+            for x in ["elapsed", "interval", "estimate", "previous"]
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class IncidentUptakeProjection(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "interval", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert (
+            self["region"].dtype == pl.Utf8
+        ), "Column 'region' and/or 'season' is not a string."
+        assert all(
+            self[x].dtype.is_numeric() for x in ["interval", "estimate"]
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self, rollout: dt.datetime) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
```

Comment:
Handling season here seems convenient for the user. Season is merely the calendar year during the autumn portion of the uptake curve. If the month is July-December, the season is the calendar year, but if the month is January-June, the season is the calendar year minus one. This is entirely dictated by the date, so why have the user worry about pre-processing the season as well as the date? As long as the date is correct, we can extract the season here.

Once season is extract, no computation is performed on it. As you say, it is merely a grouping factor, which will be important when it comes time to train on multiple seasons at once in a multilevel model.

Reply:
I'm saying that season assignment should be a matter of data preprocessing, for the code that parses & cleans the NIS data, rather than for the object itself. So I don't disagree that one should have seasons, but I think this should happen upstream of this object.

Also strong preference for season for be things like `"2024/2025"` to avoid ambiguity

Confusingly, different diseases have different "seasons" in CDC's jargon, so we might need to be careful with that in some places